lib/pbench/agent/utils.py

```diff
@@ -390,3 +393,36 @@ def collect_local_info(pbench_bin):
         hostdata[arg] = cp.stdout.strip() if cp.stdout is not None else ""
 
     return (version, seqno, sha1, hostdata)
+
+
+class LocalRemoteHost:
+    """Dynamically build a list of all the local IP addresses so that a simple
+    simple query interface provides the answer to the question, "is the given
+    host name or IP address remote?"
+    """
+
+    def __init__(self):
+        """Build up a list of local IP addresses from all the interfaces.
+        """
+        local_ips_l = []
+        for ifaceName in interfaces():
+            for ifaddr in ifaddresses(ifaceName).setdefault(AF_INET, [{"addr": None}]):
+                ip_addr = ifaddr["addr"]
+                if ip_addr is None:
+                    continue
+                local_ips_l.append(ip_addr)
+        self.local_ips = frozenset(local_ips_l)
+
+    def is_remote(self, host_name):
+        """Returns True if the given host name (or IP address) is remote else
+        returns False.
+        """
+        try:
+            ip_addr = socket.gethostbyname(host_name)
```

Comment:
`socket.gethostbyname` returns a *single* IPv4 address, according to the documentation. To get IPv6 as well, or multiple addresses, you apparently need to use `socket.getaddrinfo` instead.

I'm also trying to figure out what's documented, if anything, about the format of an IPv6 address reported here; the documentation examples I've seen are all in full 8-group format without an shortening, but I'm pretty sure from other contexts that it can return shortened names (e.g., `::1` for `localhost`). And, if so, we need to be careful about tests like `ip_address in self.local_ips`, because `0:0:0:0:0:0:0:1` and `0000:0000:0000:0000:0000:0000:0000:0001` need to match `::1` as well as `::0000:0:1` and other forms.

Python's `ipaddress` package looks a lot like the Java version I've used before: `ipaddress.ip_address(address)` is a factory creating either an `IPv4Address` object or an `IPv6Address` object depending on the value, and the objects have comparison operators so `x == a` works regardless of the original format. It's possible that your `netifaces` can do that, too, or tie into it. Anyway, the simple solution would be to feed each IP address into `ipaddress.ip_address`, store the list of resulting objects (`IPv4Address` and `IPv6Address`), the convert the input parameter as well and compare those rather than the raw text.

What you've got here will work only if you get lucky, and someone supplies the exact same form you got back from `netifaces`. If you want to explicitly say "we're not going to worry about IPv6 addresses for this spin", I won't object; but we'll definitely have to deal with that at some point.

Reply:
Hmmm...what I think we're trying to determine is whether sending a packet to `host_name` will use local transport or whether it will go out through a NIC; and, to do that we are trying to match "the address" for `host_name` to one of the addresses for the local host.  It's not clear to me whether the fact that there might be multiple address translations for `host_name` makes that harder or not -- it seems likely to me that if any of the translations are non-local then they'll all be non-local, so any one of them is as good for our purposes as the others.  (Of course, I could just be confused....)

However, the problem of comparing IPv6 addresses is very real:  they need to be cast into a canonical format before they can be compared.  If `ipaddress` would handle that for us, that would be good.

Reply:
In an ideal universe (is this one of those?), all addresses for the local host will be local. However, in an IPv6-only network, `gethostbyname` is probably useless.

Reply:
But, in this usage, we're applying `gethostbyname()` to the _target_ address, not the local host.  (We created the list of local addresses in the c'tor.)  So, assuming that the target's addresses are either all local or all remote, it should work to arbitrarily use one of them to compare against the list of local addresses (modulo, obviously, address format issues).

Reply:
As long as the target host *has* an IPv4 address, sure; but eventually we'll have to deal with an environment where it doesn't, and `getaddrinfo` avoids that issue by supporting both from the beginning. You're correct that we should be able to take any one of the addresses it returns since we have all `localhost` aliases.

Reply:
> As long as the target host has an IPv4 address

OK, I'd neglected the fact that `gethostbyname()` [only returns IPv4 addresses](https://docs.python.org/3/library/socket.html#socket.gethostbyname).  _[Le sigh]_  So, yeah, @portante, you should consider using [`getaddrinfo()`](https://docs.python.org/3/library/socket.html#socket.getaddrinfo), instead.  (And, we need to confirm that what `ifaddresses()` returns for the local host matches what `getaddrinfo()` returns for it if it has IPv6 configured, or we have to convert the formats.)

Reply:
Addressed in the new implmentation.